File: hello-dyn-kotlin/src/main/kotlin/CalculatorImplementations.kt
import kotlin.math.pow

// 基础计算器实现
class BasicCalculator : Calculator {
    override fun addFn(a: Double, b: Double): Double = a + b

    override fun subFn(a: Double, b: Double): Double = a - b

    override fun mulFn(a: Double, b: Double): Double = a * b

    override fun divFn(a: Double, b: Double): Double {
        if (b == 0.0) {
            throw CalculatorException("除数不能为零")
        }
        return a / b
    }
}

// 高级计算器实现
class AdvancedCalculator : Calculator {
    override fun addFn(a: Double, b: Double): Double {
        println("[ADVANCED] 加法: ${"%.2f".format(a)} + ${"%.2f".format(b)}")
        return a + b
    }

    override fun subFn(a: Double, b: Double): Double {
        println("[ADVANCED] 减法: ${"%.2f".format(a)} - ${"%.2f".format(b)}")
        return a - b
    }

    override fun mulFn(a: Double, b: Double): Double {
        println("[ADVANCED] 乘法: ${"%.2f".format(a)} * ${"%.2f".format(b)}")
        return a * b
    }

    override fun divFn(a: Double, b: Double): Double {
        println("[ADVANCED] 除法: ${"%.2f".format(a)} / ${"%.2f".format(b)}")
        if (b == 0.0) {
            throw CalculatorException("除数不能为零")
        }
        return a / b
    }

    // 高级功能：幂运算
    fun powFn(base: Double, exponent: Double): Double {
        println("[ADVANCED] 幂运算: ${"%.2f".format(base)} ^ ${"%.2f".format(exponent)}")
        return base.pow(exponent)
    }
}

// 科学计算器实现
class ScientificCalculator : Calculator {
    override fun addFn(a: Double, b: Double): Double {
        println("[SCIENTIFIC] 加法: ${"%.2f".format(a)} + ${"%.2f".format(b)}")
        return a + b
    }

    override fun subFn(a: Double, b: Double): Double {
        println("[SCIENTIFIC] 减法: ${"%.2f".format(a)} - ${"%.2f".format(b)}")
        return a - b
    }

    override fun mulFn(a: Double, b: Double): Double {
        println("[SCIENTIFIC] 乘法: ${"%.2f".format(a)} * ${"%.2f".format(b)}")
        return a * b
    }

    override fun divFn(a: Double, b: Double): Double {
        println("[SCIENTIFIC] 除法: ${"%.2f".format(a)} / ${"%.2f".format(b)}")
        if (b == 0.0) {
            throw CalculatorException("除数不能为零")
        }
        return a / b
    }

    // 科学功能
    fun sin(x: Double): Double = kotlin.math.sin(x)
    fun cos(x: Double): Double = kotlin.math.cos(x)
    fun tan(x: Double): Double = kotlin.math.tan(x)
    fun log(x: Double): Double = kotlin.math.ln(x)
    fun exp(x: Double): Double = kotlin.math.exp(x)
}

File: hello-dyn-dart/lib/calculator_implementations.dart
import 'calculator.dart';
import 'dart:math' as math;

// 基础计算器实现
class BasicCalculator implements Calculator {
  @override
  double addFn(double a, double b) {
    return a + b;
  }

  @override
  double subFn(double a, double b) {
    return a - b;
  }

  @override
  double mulFn(double a, double b) {
    return a * b;
  }

  @override
  double divFn(double a, double b) {
    if (b == 0) {
      throw const CalculatorException('除数不能为零');
    }
    return a / b;
  }
}

// 高级计算器实现
class AdvancedCalculator implements Calculator {
  @override
  double addFn(double a, double b) {
    print('[ADVANCED] 加法: ${a.toStringAsFixed(2)} + ${b.toStringAsFixed(2)}');
    return a + b;
  }

  @override
  double subFn(double a, double b) {
    print('[ADVANCED] 减法: ${a.toStringAsFixed(2)} - ${b.toStringAsFixed(2)}');
    return a - b;
  }

  @override
  double mulFn(double a, double b) {
    print('[ADVANCED] 乘法: ${a.toStringAsFixed(2)} * ${b.toStringAsFixed(2)}');
    return a * b;
  }

  @override
  double divFn(double a, double b) {
    if (b == 0) {
      throw const CalculatorException('除数不能为零');
    }
    print('[ADVANCED] 除法: ${a.toStringAsFixed(2)} / ${b.toStringAsFixed(2)}');
    return a / b;
  }

  // 额外功能
  double powFn(double base, double exponent) {
    return math.pow(base, exponent).toDouble();
  }
}
